src/folding/nova/circuits.rs

```diff
@@ -0,0 +1,282 @@
+pub struct CommittedInstanceE2Var<C: CurveGroup, GC: CurveVar<C, CF2<C>>>
+where
+    for<'a> &'a GC: GroupOpsBounds<'a, C, GC>,
+{
+    _c: PhantomData<C>,
+    cmE: GC,
+    cmW: GC,
+}
+
+impl<C, GC> AllocVar<CommittedInstance<C>, CF2<C>> for CommittedInstanceE2Var<C, GC>
+where
+    C: CurveGroup,
+    GC: CurveVar<C, CF2<C>>,
+    for<'a> &'a GC: GroupOpsBounds<'a, C, GC>,
+{
+    fn new_variable<T: Borrow<CommittedInstance<C>>>(
+        cs: impl Into<Namespace<CF2<C>>>,
+        f: impl FnOnce() -> Result<T, SynthesisError>,
+        mode: AllocationMode,
+    ) -> Result<Self, SynthesisError> {
+        f().and_then(|val| {
+            let cs = cs.into();
+
+            let cmE = GC::new_variable(cs.clone(), || Ok(val.borrow().cmE), mode)?;
+            let cmW = GC::new_variable(cs.clone(), || Ok(val.borrow().cmW), mode)?;
+
+            Ok(Self {
+                _c: PhantomData,
+                cmE,
+                cmW,
+            })
+        })
+    }
+}
+
+/// Implements the circuit that does the checks of the Non-Interactive Folding Scheme Verifier
+/// described in section 4 of https://eprint.iacr.org/2021/370.pdf, where the cmE & cmW checks are
+/// delegated to the NIFSCycleFoldGadget.
+pub struct NIFSGadget<C: CurveGroup> {
+    _c: PhantomData<C>,
+}
+
+impl<C: CurveGroup> NIFSGadget<C>
+where
+    C: CurveGroup,
+{
+    /// Implements the constraints for NIFS.V for u and x, since cm(E) and cm(W) are delegated to
+    /// the CycleFold circuit.
+    pub fn verify(
+        r: FpVar<CF1<C>>,
+        ci1: CommittedInstanceE1Var<C>,
+        ci2: CommittedInstanceE1Var<C>,
+        ci3: CommittedInstanceE1Var<C>,
+    ) -> Result<(), SynthesisError> {
+        ci3.u.enforce_equal(&(ci1.u + r.clone() * ci2.u))?;
+
+        let x_rlc = ci1
+            .x
+            .iter()
+            .zip(ci2.x)
+            .map(|(a, b)| a + &r * &b)
+            .collect::<Vec<FpVar<CF1<C>>>>();
+        x_rlc.enforce_equal(&ci3.x)?;
+
+        Ok(())
+    }
+}
+
+/// NIFSCycleFoldGadget performs the Nova NIFS.V elliptic curve points relation checks in the other
+/// curve following CycleFold (https://eprint.iacr.org/2023/1192.pdf).
+pub struct NIFSCycleFoldGadget<C: CurveGroup, GC: CurveVar<C, CF2<C>>> {
+    _c: PhantomData<C>,
+    _gc: PhantomData<GC>,
+}
+impl<C: CurveGroup, GC: CurveVar<C, CF2<C>>> NIFSCycleFoldGadget<C, GC>
+where
+    C: CurveGroup,
+    GC: CurveVar<C, CF2<C>>,
+    for<'a> &'a GC: GroupOpsBounds<'a, C, GC>,
+{
+    pub fn verify(
+        r_bits: Vec<Boolean<CF2<C>>>,
+        r2_bits: Vec<Boolean<CF2<C>>>, // r^2
+        cmT: GC,
+        ci1: CommittedInstanceE2Var<C, GC>,
+        ci2: CommittedInstanceE2Var<C, GC>,
+        ci3: CommittedInstanceE2Var<C, GC>,
+    ) -> Result<(), SynthesisError> {
+        // cm(E) check:
+        ci3.cmE.enforce_equal(
+            &(ci1.cmE
+                + cmT.scalar_mul_le(r_bits.iter())?
+                + ci2.cmE.scalar_mul_le(r2_bits.iter())?),
+        )?;
```

Comment:
If we scale `ci2.cmE` twice by `r_bits` we get the same result, and we could avoid having `r2_bits`, which might be dense even `r_bits` is not so dense (e.g. in Nova it's 128-bits).

```suggestion
        ci3.cmE.enforce_equal(
            &((ci2.cmE.scalar_mul_le(r_bits.iter())? + cmT).scalar_mul_le(r_bits.iter())?
                + ci1.cmE),
        )?;
```

Reply:
That's a very good point, thanks! ^^
Updated in the commit.
Addionally I like it because avoids us from computing `r^2` plus converting it into bits in-circuit, since in-circuit we only need it in bits format for the point scalar muls.

I've checked it and for `r` of `128 bits` in pasta curves, the circuit with `r2_bits` takes `5198 constraints`, and the one using 2 scalar muls with `r_bits` takes `5179 constraints` (and if we also substract the constraints for having `r2_bits` as input, it takes then `4923 constraints`). The same test with `r` of `250 bits`, takes `7516 constr` for the `r2_bits` approach, and `9693 constr` for the 2 scalarmuls by `r_bits` approach. So the number of constraints gets reduced for 128 bits as you said.